Drag & drop multiple tree nodes in structure tree

thomaslow opened the pull request:
Fixes #5926 

This pull request adds support for dragging multiple tree node elements in the structure trees.

This feature requires that multiple tree nodes can be selected in the structure trees (via `ctrl + click` or `shift + click`). Primefaces already supports this via the `selectionMode="multiple"` setting. However, there are a lot of side effects of this change. Although the metadata editor backend was already capable of multi-selection in the gallery, it was only designed to select multiple media elements, but not multiple logical structure elements. Also, Primefaces does not support to distinguish the last selected element (solid border in the gallery) from the previously selected elements (dashed border in the gallery).

As a consequence, there are features in the metadata editor that are currently not designed to work intuitively when selecting multiple tree nodes:

- the metadata panel doesn't know what to show when multiple elements are selected
- a lot of context menu items are not optimized for multiple selected elements, e.g. to open media in a new window, or to delete multiple selected elements
- the gallery does not allow to select multiple logical structures (stripes)

I already fixed some of these issues, even though they are technically not related to the "drag & drop" feature:

- the selection of media in the physical structure tree (separate media mode) did not work in the current master branch
- the "upload media" context menu is now hidden if multiple elements are selected, the dialog assumes a single-selection
- the add structure dialog now accepts multiple selections from the structure tree, which previously only worked when selecting multiple media in the gallery

I think there are still a lot more features that can be improved now, but are out of the scope of this issue and pull request.

A demonstration video of the multi drag-and-drop feature:

https://github.com/user-attachments/assets/8c63c6d7-7de5-4e2e-b5d7-c52fed89bbfa

Todo
- [x] Update Javascript to synchronize selection state to gallery and vertical thumbnail list
- [x] Improve performance for large trees
- [x] Also update physical tree
- [x] Improve code style
- [x] Add selenium tests
- [x] Fix problem that gallery selection is cleared after drag & drop
- [x] Backport to 3.8.x

~~@solth Let me know if you think this conflicts too much with your current development plans (PrimeFaces update) or whether this could be merged sooner.~~

solth commented:
> @solth Let me know if you think this conflicts too much with your current development plans (PrimeFaces update) or whether this could be merged sooner.

No, judging by the current set of changes in this pull request I think it is no problem to merge this pull request before the PrimeFaces update. I am surprised that the multi select in the tree can be implemented with our current PrimeFaces version 8, though. I thought there was a bug in PrimeFaces 8 with the `row_index` value that prevented us from resolving #5926 before the update, but I am unable to find that bug report anymore.

solth commented:
This also fixes #3629

solth reviewed (commented):
@thomaslow thanks a lot for refactoring and cleaning up a lot of the code in the metadata editor and also adding some much needed comments to better understand some complex parts of the code. 

The multi selection of tree nodes works very well. Code-wise I just a have few minor suggestions and typo fixes. One observation I made was that selections remain after drag'n'drop actions in the tree, but are cleared in the gallery:

<img width="1500" alt="Bildschirmfoto 2025-02-14 um 14 37 49" src="https://github.com/user-attachments/assets/e6fd9733-acfb-4784-8a27-0956037fdc89" />

I did clear my browser cache before testing to make sure to use the update Javascript and CSS files, so there seems to be some minor problem with the synchronization here.

oliver-stoehr reviewed (approved):
Thank you for this great addition! This is a functional review:
The multiselect in the structure tree works very well and seems to be very robust. I could not find any bug or combination of actions that should behave differently when using drag & drop in the structure tree.
The synchronization with the other panels works as expected.

BartChris commented:
I also did some testing and so far it worked good, great work! Would it be possible to consider it for backport to 3.8 to roll it out to the users in a shorter time frame?

solth commented:
> I also did some testing and so far it worked good, great work! Would it be possible to consider it for backport to 3.8 to roll it out to the users in a shorter time frame?

I wanted to suggest the same thing, if @thomaslow thinks a backport can be provided without too much effort. Even though the changes are substantial, I think these improvements are so valuable that they could and probably should be included in a bugfix release. (one could argue the missing support for multiselect for trees was a bug, after all!)

thomaslow commented on Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/AddDocStrucTypeDialog.java:
```diff
@@ -181,13 +182,14 @@ private void addSingleDocStruc(boolean selectViews) {
+            } catch (NoSuchMetadataFieldException e) {
+                Helper.setErrorMessage(e.getLocalizedMessage(), logger, e);
             }
+            dataEditor.refreshStructurePanel();
```
You are right. The one after updating the selection is not required. I fixed it.

thomaslow commented on Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/StructureTreeOperations.java:
```diff
@@ -0,0 +1,267 @@
+ */
+public class StructureTreeOperations {
+
+    private static final Logger logger = LogManager.getLogger(StructureTreeOperations.class);
```
Removed it.

thomaslow commented on Kitodo/src/main/webapp/WEB-INF/resources/js/metadata_editor.js:
```diff
@@ -847,21 +973,101 @@ metadataEditor.physicalTree = {
+     * 
+     * @returns the selected nodes
+     */
+    _findSelectedNodes() {
```
This is a bit tricky. I didn't want to repeat the code of the tree node click handler both for the logical and physical tree. That's why I use the statement `metadataEditor.logicalTree.onNodeClick.apply(this, [node, event]);` in line 959 to call the `onNodeClick` function with a different `this` (meaning `metadataEditor.physicalTree`).

Eventually, we should transform the code to more modern Javascript and use classes and class inheritance to avoid writing the same code for both the logical and physical structure tree.

thomaslow commented on Kitodo/src/main/webapp/WEB-INF/resources/js/metadata_editor.js:
```diff
@@ -847,21 +973,101 @@ metadataEditor.physicalTree = {
+     * @param node jquery object of a tree node
+     * @returns true if the node is a media node
+     */
+    _isMediaNode(node) {
```
That function is actually not used any more. I removed it.

thomaslow commented:
@solth I fixed the problem that the gallery selection is lost after drag & drop in the structure tree.

Unfortunately, I discovered two new issues that can not easily be resolved except for updating Primefaces to a newer version. 

Both problems are client-side Javascript issues. 

The first issue prevents users from dragging more than the first 10 nodes of a subtree. This is already fixed in Primefaces v13.0.5:
https://github.com/primefaces/primefaces/issues/11252
I actually tested this scenario with an even larger selection. It is only a problem when moving two nodes with a row key number that is a prefix of the second row key (e.g. moving node `1_1` and node `1_11`). Moving more than 10 nodes works fine when they all have the same number of digits (e.g. moving node `1_34` to `1_84`).

A demonstration of the problem:

https://github.com/user-attachments/assets/7f438fe8-6125-4af0-9f78-89981fca3fb4

The other issue can be reproduced by selecting two folders with one folder in between. When dragging these nodes into the in-between folder, the Primefaces Javascript silently stops in the middle of the drag and drop operation, partially updates the user interface, but does not send a request to the backend at all. As a consequence, the user interface tree and backend tree are not synchronized any more. Afterwards, there can be various issues, e.g. seemingly random drag & drop behavior, or a null pointer exception as demonstrated in this video:

https://github.com/user-attachments/assets/deff15eb-6ce3-4739-8e64-d691a67c1878

This only happens if you drop the current selection one node above a selected node while having another node selected above the drop point. Unfortunately, I did not find a way to prevent this. We would have to modify the Primefaces Javascript drag and drop logic. Also, I didn't find a bug report for this. Hopefully it is fixed in a newer version of Primefaces.

Let me know whether you think these issues are critical and will prevent the pull request from being merged.

thomaslow commented on Kitodo/src/main/java/org/kitodo/production/forms/dataeditor/PaginationPanel.java:
```diff
@@ -363,24 +360,11 @@ public void startPaginationClick() {
         preparePaginationSelectionItems();
         dataEditor.refreshStructurePanel();
-        updateMetadataPanel();
+        dataEditor.updateToDefaultSelection();
```
Yes, it makes more sense that the selection is preserved here. I fixed it.

thomaslow commented:
@solth I fixed the `dragAndDropPageTest` such that the original page order (2-1-3) is restored and saved at the end of the test. And, I updated the `movePagesToUnsavedStructureTest` such that it assumes the original page order (2-1-3) and can now be executed independently from the `dragAndDropPageTest`.

solth commented:
> Let me know whether you think these issues are critical and will prevent the pull request from being merged.

@thomaslow thank you for the update. I think the first issue you mentioned can be seen as merely an inconvenience. Not being able to move more than 10 pages may be a hassle, but judging by your example video it will at least be very clear to users that some pages have not been moved. Ideally we should prevent people from marking more pages, in my option, or show a warning dialog to inform them about the issue, but that would be optional in my opinion and the additional functionality and improvements outweigh this problem, IMHO.

The second issue seems more critical to me, though. Being confronted with random behaviour after the interaction you described will probably lead to much frustration and many bug reports, especially if there is a chance to encounter unhandled exceptions and potentially lose unsaved work. At the moment I think that problem is indeed a dealbreaker and I wouldn't want to introduce it into the main branch, unless we find a way to prevent the user from performing this problematic UI action (e.g. dragging multiple objects onto a structure _between_ those objects). Do you think that would at all be feasible?

solth commented:
> @solth I fixed the `dragAndDropPageTest` such that the original page order (2-1-3) is restored and saved at the end of the test. And, I updated the `movePagesToUnsavedStructureTest` such that it assumes the original page order (2-1-3) and can now be executed independently from the `dragAndDropPageTest`.

Thanks a lot for refactoring these tests!

thomaslow commented:
> At the moment I think that problem is indeed a dealbreaker and I wouldn't want to introduce it into the main branch, unless we find a way to prevent the user from performing this problematic UI action (e.g. dragging multiple objects onto a structure _between_ those objects). Do you think that would at all be feasible?

@solth I would have to look into it. However, my first impression is, that preventing this issue (e.g. by restricting users from performing this specific drag-and-drop interaction) would be just as difficult as fixing the problem itself, and also require injecting additional Javascript code that strictly depends on the current PrimeFaces version. A small patch of the PrimeFaces Javascript code was already necessary, see [metadataeditor_tree_drag_drop.js](https://github.com/kitodo/kitodo-production/blob/f9409b440a6a30180325da9431b9ebe5b46fae9a/Kitodo/src/main/webapp/WEB-INF/resources/js/metadataeditor_tree_drag_drop.js).

If you think it would be acceptable to copy, modify and inject large portions of the Primefaces Tree Javascript code, I could give this a try. Essentially, I would try to overwrite some of the Javascript methods with their more recent versions, see [tree.vertical.widget.js](https://github.com/primefaces/primefaces/blob/master/primefaces/src/main/frontend/packages/components/src/tree/tree.vertical.widget.js). Hopefully, after upgrading Primefaces, we could remove this "backport".

In any case, the current Primefaces 14 showcase does not allow to test this specific drag-and-drop scenario. I do not even know, whether the problem is fixed in the most recent version.

solth commented:
@thomaslow after discussing this with @subhhwendt I would suggest to deactivate multi node drag'n'drop with `multipleDrag="false"` in the structure tree (and also set the corresponding tests to `disabled` with an explanation about the bug) until the PrimeFaces update has been completed. At that point you can check whether the bug has been fixed in the new PrimeFaces version and if just re-enabling the `multidrag` option in the tree is enough to resolve #5926 at that point (which we anticipated to be implemented after the PrimeFaces update anyway).

If we proceed that way we can merge _this_ pull request right now which would already resolve #3629 and add the other valuable refactorings you made to the code while avoiding the critical bug you discovered. And we could avoid keeping this pull request up-to-date with the changes introduced into the `main` branch in the meantime. If you agree, I would then unlink #5926 from this PR and merge it once the `multidrag` has been deactivated.

thomaslow commented:
@solth I disabled the `multidrag` option, split the selenium tests into two test cases and disabled the one for multi-drag and drop. Now, when multiple tree nodes are selected, only the actual dragged node is moved. 

I tested that the multi-selection and single-drag and drop fundamentally works. However, I would appreciate someone double-checking the current partially disabled version again before merging.

oliver-stoehr reviewed (approved):
I tested the functionality with the latest version and think it's fine. Overall our drag & drop in Kitodo is not the most intuitive compared to other web applications, but these changes work as intended and I still find the drag & drop smooth enough to be merged.
The fact that drag & drop operations behave differently despite the same selection in the structure panel and gallery is, in my opinion, relatively easy to understand and works quite well once you've tried it one or two times.

BartChris commented:
It works good for me. I do not know how much work is required to allow discontinous shift selections as described in my comment. The use case this came up for was pagination which should in the future mostly be based on mutiselect in the structure tree. (See https://github.com/kitodo/kitodo-production/issues/6455) 
Implementing that seems to require some work. Maybe this can be seen as an future enhancement which probably also has to be discussed. (What behaviour of shift+select do people want?).

thomaslow commented on Kitodo/src/main/webapp/WEB-INF/resources/js/metadata_editor.js:
```diff
@@ -744,26 +746,71 @@ metadataEditor.logicalTree = {
+        let currentTreeNodeId = this._getTreeNodeIdFromNode(node);
+
+        if (!(event.metaKey || event.ctrlKey)) {
+            // remove all previous selections when clicking without ctrl key or shift key
```
Yes, the comment is misleading here. I think I experimented with the selection modifier keys (ctrl, shift), but later realized that PrimeFaces does not support multiple shift selections or shift selections over multiple hierarchy levels.

thomaslow commented:
@solth Thanks for merging. Please let me know in case I can help with any significant merge conflicts.

solth commented:
@thomaslow thanks for the offer. I am indeed struggeling a little bit with the conflicts in `StructurePanel.java` in the context of #6482 right now, but your pull request has been waiting much longer for a merge so I just found it fair to merge it first. 

If you don't mind, I would merge the corresponding pull request against the 3.8.x branch in a different order, though, so first #6483, then #6440. Then the task of resolving the conflicts would fall on you, but I presume that will be much easier for you since the largest part of changes comes from you anyway. Is that ok for you?

thomaslow commented:
@solth I think it would be more efficient if you/we merge both pull requests in the same order in both branches, otherwise - I believe - we will have different merge conflicts and have to do the same work twice.

 I can have a look at #6482 today, either helping with merge conflicts or reviewing / testing.

solth commented:
> I can have a look at #6482 today, either helping with merge conflicts or reviewing / testing.

Thanks. I finished resolving the conflicts and it seems large parts of my custom logic required to find nodes corresponding to `Views` selected in the gallery when media is hidden in the structure tree is not required anymore thanks to your code changes in _this_ pull request. I will request an additional review from you in #6482 for you to check if my assumption is correct.
